### DISCUSSION TITLE
Feedback wanted 👋🏼

### DISCUSSION THREAD

**Discussion by ahx** [ahx/openapi_first#169] [General]
I am wondering if anyone is using this and what they like and dislike about this tool. I have some ideas what to change, but I'd like to hear from people outside. Feel free to create issues or reply here.

**Comment by twe4ked** [comment]
Hey, I really like the look of this gem, thanks!

One thing I'd like to see with the structure of the gem is some more separation between the middlewares/router and the underlying logic. Ideally I'd like a simple API that I can pass a request/response to and have it raise exceptions for invalid responses so I can easily build my own middleware. See #140 for a bit more detail.

Thanks :)

**Comment by MrBananaLord** [comment]
Hi! Thanks for creating this gem!

I have pretty similar feedback. Having ready-to-use middleware is really nice but it comes with some limitations. I would prefer to have the option of directly using the validation as described here (https://github.com/ahx/openapi_first/issues/140#issuecomment-1299355878).

I would be happy to contribute if you'd be willing to go that direction :)

**Comment by ahx** [comment]
Hi @MrBananaLord, if people think that use case is useful in their situation I am happy to support that. Instead of raising an exception by default I would rather return a [`ValidationResult`](https://github.com/ahx/openapi_first/blob/main/lib/openapi_first/validation_result.rb) instance or something similar. I think having access to all details why the validation failed would be a good start to put custom error handling on top. I am adding custom error responses in 1.0 and I think that would be a good opportunity to implement [Manual request/response validation](https://github.com/ahx/openapi_first/issues/195). So here is an [issue about that](https://github.com/ahx/openapi_first/issues/195). I am happy to receive your feedback there.

**Comment by ahx** [comment]
Thanks for your feedback on this. Here is another discussion that is related to what has been said here. https://github.com/ahx/openapi_first/discussions/221 I would love to get your feedback.
